content/post/date_features.Rmd

```diff
@@ -0,0 +1,184 @@
+```{r}
+tsk_train = tsk$clone()$filter(train.idx)
+tsk_test = tsk$clone()$filter(test.idx)
+```
+
+To estimate the performance on future data, we will use `5-fold cross-validation`:
+```{r}
+cv5 = rsmp("cv", folds = 5)
+```
+
+As our baseline model, we use a random forest, `r ref("mlr_learners_regr.ranger", "ranger learner")`.
+For the baseline, we only use the original features that are sensible and drop `instant` (record
+index), `dteday` (year-month-day as a `character`, not usable), `date` (our new `POSIXct` variable
+which we will only use later), `casual` (count of casual users) and `registered` (count of
+registered users). Note that `casual` and `registered` adds up to `cnt`.
+```{r}
+lrn_rf = lrn("regr.ranger")
+tsk_train_rf = tsk_train$clone()$select(setdiff(tsk$feature_names,
+  c("instant", "dteday", "date", "casual", "registered")))
+```
+
+We can then use `resample` with `5-fold cross-validation`:
+```{r}
+res_rf = resample(tsk_train_rf, learner = lrn_rf, resampling = cv5)
+res_rf$score(msr("regr.mse"))
+```
+
+The average RMSE is given by:
+```{r}
+sprintf("RMSE ranger original features: %s", round(sqrt(res_rf$aggregate()), digits = 2))
+```
+
+We now want to improve our baseline model by using newly engineered features based on the `date`
+`POSIXct` column.
+
+## PipeOpDateFeatures
+To engineer new features we use `PipeOpDateFeatures`. This pipeline automatically dispatches on
+`POSIXct` columns of the data and by default adds plenty of new date-time related features (for
+details, see `r help(PipeOpDateFeatures)`). Here, we want to add all expect for `minute` and
+`second`, because this information is not available. As we additionally want to use cyclical
+versions of the features we set `cyclic = TRUE`:
+```{r}
+library(mlr3pipelines)
+pop = po("datefeatures", param_vals = list(cyclic = TRUE, minute = FALSE, second = FALSE))
+```
+
+Training this pipeline will result in simply adding the new features (and removing the original
+`POSIXct` feature(s) used for the feature engineering, see also the `keep_date_var` parameter). In
+our training task, we can now also drop the features, `yr`, `mnth`, `hr`, and `weekday`, because our
+pipeline will generate these anyways:
+```{r}
+tsk_train_ex = tsk_train$clone()$select(setdiff(tsk$feature_names,
+  c("instant", "dteday", "yr", "mnth", "hr", "weekday", "casual", "registered")))
+pop$train(list(tsk_train_ex))
+```
+
+To combine this feature engineering step with a random forest,
+`r ref("mlr_learners_regr.ranger", "ranger learner")`, we now construct a `GraphLearner`.
+
+## Using the New Features in a GraphLearner
+We create a `GraphLearner` consisting of the `PipeOpDateFeatures` pipeline and a
+`r ref("mlr_learners_regr.ranger", "ranger learner")`. This `GraphLearner` then behaves like any
+other `Learner`:
+```{r}
+grl = GraphLearner$new(
+  po("datefeatures", param_vals = list(cyclic = TRUE, minute = FALSE, second = FALSE)) %>>%
+  po("learner", lrn("regr.ranger"))
+)
+```
+
```

Comment:
Maybe also mention or use `removeconstants` PO here. 
Reason: Many algos can not deal with constant features and therefore this might be required.